### ISSUE TITLE
Get rid of `template<t = T>` in atomics

### ISSUE THREAD

**Issue by mrzv** [illuhad/hipSYCL#211]
I mentioned this in the PR discussion, but I think it got lost. Several functions in `atomic` use the following construction:
```
  template<class t = T,
           HIPSYCL_CONDITIONALLY_ENABLE_ATOMICS(t)>
  HIPSYCL_KERNEL_TARGET
  t fetch_add(t operand, memory_order memoryOrder =
      memory_order::relaxed) volatile
```
This causes a problem in the following example:
```
atomic<unsigned long long> x = ...;
unsigned long long y = x.fetch_add(1);
```
`y` gets only the lowest 32-bits of `x` because the argument 1 is interpreted as an integer. The fix is to pass `1LU`, but I find this very non-obvious behavior and a tricky thing to debug.

My suggestion is to get rid of `t = T` entire and just keep:
```
  template<HIPSYCL_CONDITIONALLY_ENABLE_ATOMICS(T)>
```

**Comment by psalz** [comment]
The `t` is required for SFINAE (`T` is substituted at the class-level, before the member function is looked up, whereas SFINAE on a member function only works if the substitution fails during overload resolution). To avoid having `t` and `T` being different types, maybe you could try adding something like a `static_assert(std::is_same_v<t, T>)`..? Or it could be added as an additional SFINAE condition.

**Comment by illuhad** [comment]
In principle we could drop SFINAE entirely in this case and just do something like `static_assert(std::is_integral_v<T> || HIPSYCL_EXT_FP_ATOMICS, "Unsupported datat types for atomic operations")`. I see only one issue with that: backends may support different types for different floating point atomics, and SFINAE allows us to forward the responsibility for using fp atomics to the user: Only those functions that the user calls will be instantiated. FP atomics that are unsupported on a backend will not cause an issue unless the user actually tries to use them.

Since we're moving to C++17, maybe this can also be made more elegant with `if constexpr`?

**Comment by mrzv** [comment]
Actually, what happens, if we instead change the argument to the function to `T`:
```
template<class t = T,
           HIPSYCL_CONDITIONALLY_ENABLE_ATOMICS(t)>
  HIPSYCL_KERNEL_TARGET
  T fetch_add(T operand, memory_order memoryOrder =
      memory_order::relaxed) volatile
```
Then unless the user explicitly specifies a template argument, everything should work correctly. Or is it going to complain about not being able to infer `t`? It shouldn't, right?

**Comment by illuhad** [comment]
@mrzv I think this should work. Can you create a PR so that we can verify?

**Comment by mrzv** [comment]
@illuhad See #215

**Comment by illuhad** [comment]
Fixed by PR #215.
